modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DataNodesManager.java

```diff
@@ -0,0 +1,1527 @@
+
+                        DistributionZoneTimer scaleUpTimer = dataNodeHistoryContext.scaleUpTimer();
+                        DistributionZoneTimer scaleDownTimer = dataNodeHistoryContext.scaleDownTimer();
+
+                        DistributionZoneTimer modifiedScaleUpTimer = scaleUpTimer
+                                .modifyTimeToWait(zoneDescriptor.dataNodesAutoAdjustScaleUp());
+
+                        DistributionZoneTimer modifiedScaleDownTimer = scaleDownTimer
+                                .modifyTimeToWait(zoneDescriptor.dataNodesAutoAdjustScaleDown());
+
+                        Pair<HybridTimestamp, Set<NodeWithAttributes>> currentDataNodes = currentDataNodes(
+                                timestamp,
+                                dataNodesHistory,
+                                modifiedScaleUpTimer,
+                                modifiedScaleDownTimer,
+                                zoneDescriptor
+                        );
+
+                        DistributionZoneTimer scaleUpTimerToSave = timerToSave(timestamp, modifiedScaleUpTimer);
+                        DistributionZoneTimer scaleDownTimerToSave = timerToSave(timestamp, modifiedScaleDownTimer);
+
+                        return new LoggedIif(iif(
+                                and(
+                                        dataNodesHistoryEqualToOrNotExists(zoneId, dataNodesHistory),
+                                        and(
+                                                timerEqualToOrNotExists(zoneScaleUpTimerKey(zoneId), scaleUpTimer),
+                                                timerEqualToOrNotExists(zoneScaleDownTimerKey(zoneId), scaleDownTimer)
+                                        )
+                                ),
+                                ops(
+                                        addNewEntryToDataNodesHistory(zoneId, dataNodesHistory, currentDataNodes.getFirst(),
+                                                currentDataNodes.getSecond()),
+                                        renewTimer(zoneScaleUpTimerKey(zoneId), scaleUpTimerToSave),
+                                        renewTimer(zoneScaleDownTimerKey(zoneId), scaleDownTimerToSave)
+                                ).yield(true),
+                                ops().yield(false)
+                        ),
+                        updateHistoryLogRecord(zoneId, "distribution zone auto adjust change", dataNodesHistory, timestamp,
+                                currentDataNodes.getFirst(), currentDataNodes.getSecond(), scaleUpTimerToSave, scaleDownTimerToSave),
+                        "Failed to update data nodes history and timers on distribution zone auto adjust change [timestamp="
+                                + timestamp + "]."
+                        );
+                    });
+        }, zoneDescriptor);
+    }
+
+    private static String updateHistoryLogRecord(
```

Comment:
This method's name is misleading, please rename it to something like `createHistoryLogRecordMessage`

Reply:
removed it